Mathlib/Data/Finsupp/MonomialOrder.lean

```diff
@@ -0,0 +1,411 @@
+
+end MonomialOrder
+
+section Lex
+
+open Finsupp
+
+open scoped MonomialOrder
+
+-- The linear order on `Finsupp`s obtained by the lexicographic ordering. -/
+noncomputable instance {α N : Type*} [LinearOrder α] [OrderedCancelAddCommMonoid N] :
+    OrderedCancelAddCommMonoid (Lex (α →₀ N)) where
+  le_of_add_le_add_left a b c h := by simpa only [add_le_add_iff_left] using h
+  add_le_add_left a b h c := by simpa only [add_le_add_iff_left] using h
+
+theorem Finsupp.lex_lt_iff {α N : Type*} [LinearOrder α] [LinearOrder N] [Zero N]
+    {a b : Lex (α →₀ N)} :
+    a < b ↔ ∃ i, (∀ j, j< i → ofLex a j = ofLex b j) ∧ ofLex a i < ofLex b i :=
+    Finsupp.lex_def
+
+theorem Finsupp.lex_le_iff {α N : Type*} [LinearOrder α] [LinearOrder N] [Zero N]
+    {a b : Lex (α →₀ N)} :
+    a ≤ b ↔ a = b ∨ ∃ i, (∀ j, j< i → ofLex a j = ofLex b j) ∧ ofLex a i < ofLex b i := by
+    rw [le_iff_eq_or_lt, Finsupp.lex_lt_iff]
+
+/- -- #check Finsupp.toLex_monotone
+theorem _root_.Finsupp.toLex_monotone' {σ : Type*} [LinearOrder σ] :
+    Monotone (toLex (α := σ →₀ ℕ)) := by
+  apply Finsupp.toLex_monotone
+  intro a b h
+  rw [← (add_tsub_cancel_of_le h), toLex_add]
+  simp only [AddEquiv.refl_symm, le_add_iff_nonneg_right, ge_iff_le]
+  apply bot_le
+-/
+
+variable {σ : Type*} [LinearOrder σ]
+
+/-- The lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.lex [WellFoundedGT σ] :
+    MonomialOrder σ where
+  syn := Lex (σ →₀ ℕ)
+  locacm := Lex.linearOrderedCancelAddCommMonoid
+  toSyn := {
+    toEquiv := toLex
+    map_add' := toLex_add } -- AddEquiv.refl _ -- (AddEquiv.refl (Lex (σ →₀ ℕ))).symm
+  wf := Lex.wellFoundedLT
+  toSyn_monotone := Finsupp.toLex_monotone
+
+theorem MonomialOrder.lex_le_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≼[lex] d ↔ toLex c ≤ toLex d := Iff.rfl
+
+theorem MonomialOrder.lex_lt_iff [WellFoundedGT σ] {c d : σ →₀ ℕ} :
+    c ≺[lex] d ↔ toLex c < toLex d := Iff.rfl
+
+/-- The reverse lexicographic order on `σ →₀ ℕ`, as a `MonomialOrder` -/
+noncomputable def MonomialOrder.revLex [WellFoundedLT σ] :
+    MonomialOrder σ :=
+  MonomialOrder.lex (σ := σᵒᵈ)
+
+theorem MonomialOrder.revLex_le_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≼[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) ≤ (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+theorem MonomialOrder.revLlex_lt_iff [WellFoundedLT σ] {c d : σ →₀ ℕ} :
+    c ≺[revLex] d ↔ (toLex c : Lex (σᵒᵈ →₀ ℕ)) < (toLex d : Lex (σᵒᵈ →₀ ℕ)) :=
+  Iff.rfl
+
+end Lex
+
+section degLex
+
+/-- A type synonym to equip a type with its lexicographic order sorted by degrees. -/
+def DegLex (α : Type*) := α
+
+variable {α : Type*}
+/-- `toDegLex` is the identity function to the `DegLex` of a type.  -/
+@[match_pattern]
+def toDegLex : α ≃ DegLex α :=
+  Equiv.refl _
+
+/-- `ofDegLex` is the identity function from the `DegLex` of a type.  -/
+@[match_pattern]
+def ofDegLex : DegLex α ≃ α :=
+  Equiv.refl _
```

Comment:
Here, I would suggest defining only one equivalence `equivDegLex : α ≃ DegLex α`, and two functions `toDegLex : α → DegLex α` and `ofDegLex : DegLex α → α`.

Reply:
I was basically copying what is already done for `Lex`, which defines the two equivalences.